src/backend/distributed/operations/stage_protocol.c

```diff
@@ -858,6 +879,146 @@ UpdateShardStatistics(int64 shardId)
+static void
+UpdateTableStatistics(Oid relationId)
+{
+	List *citusTableIds = NIL;
+	citusTableIds = lappend_oid(citusTableIds, relationId);
+
+	/* we also want shard min/max values for append distributed tables */
+	List *connectionList = GetShardStatisticsInParallel(citusTableIds, true);
+
+	UpdateShardNameAndSizeAndMinMax(connectionList);
+}
+
+
+/*
+ * UpdateShardNameAndSizeAndMinMax receives shard name, size
+ * and min max results from the given connection list, and updates
+ * respective entries in pg_dist_placement and pg_dist_shard
+ */
+static void
+UpdateShardNameAndSizeAndMinMax(List *connectionList)
+{
+	MultiConnection *connection = NULL;
+
+	/*
+	 * From the connection list, we will not get all the shards, but
+	 * all the placements. We use a hash table to remember already visited shard ids
+	 * since we update all the different placements of a shard id at once.
+	 * Allocate sufficient capacity for O(1) expected look-up time.
+	 */
+	int capacity = 512; /* we need something meaningful here */
+	int flags = HASH_ELEM | HASH_CONTEXT | HASH_BLOBS;
+	HASHCTL info = {
+		.keysize = sizeof(uint64),
+		.entrysize = sizeof(uint64),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *alreadyVisitedShardPlacements = hash_create("alreadyVisitedShardPlacements",
+													  capacity, &info, flags);
+
+	foreach_ptr(connection, connectionList)
+	{
+		bool raiseInterrupts = true;
+
+		if (PQstatus(connection->pgConn) != CONNECTION_OK)
+		{
+			continue;
+		}
+
+		PGresult *result = GetRemoteCommandResult(connection, raiseInterrupts);
+		if (!IsResponseOK(result))
+		{
+			ReportResultError(connection, result, WARNING);
+			continue;
+		}
+
+		int64 rowCount = PQntuples(result);
+		int64 colCount = PQnfields(result);
+
+		/* Although it is not expected */
+		if (colCount != SHARD_SIZES_MIN_MAX_COLUMN_COUNT)
+		{
+			ereport(WARNING, (errmsg("unexpected number of columns from "
+									 "citus_update_table_statistics")));
+			continue;
+		}
+
+		for (int64 rowIndex = 0; rowIndex < rowCount; rowIndex++)
+		{
+			bool foundInSet = false;
+			uint64 shardId = ParseIntField(result, rowIndex, 0);
+
+			/* check for the dummy entries we put so that UNION ALL wouldn't complain */
+			if (shardId != INVALID_SHARD_ID)
+			{
+				hash_search(alreadyVisitedShardPlacements,
+							&shardId, HASH_ENTER,
+							&foundInSet);
+				if (foundInSet)
+				{
+					/* We have already updated this placement list */
+					continue;
+				}
+
+				char *minValueResult = PQgetvalue(result, rowIndex, 1);
+				char *maxValueResult = PQgetvalue(result, rowIndex, 2);
+				text *shardMinValue = cstring_to_text(minValueResult);
+				text *shardMaxValue = cstring_to_text(maxValueResult);
+				uint64 shardSize = ParseIntField(result, rowIndex, 3);
+				ShardInterval *shardInterval = LoadShardInterval(shardId);
+				Oid relationId = shardInterval->relationId;
+				char storageType = shardInterval->storageType;
+				List *shardPlacementList = ActiveShardPlacementList(shardId);
```

Comment:
After this part, we seem to have some duplicate with `UpdateShardStatistics`. Have you thought about using a method for removing this duplication?

Reply:
Good idea, will do that.